drivers/gpu/drm/rp1/rp1-vec/rp1_vec_hw.c

```diff
@@ -46,334 +53,371 @@ static const struct rp1vec_ipixfmt my_formats[] = {
+	 * would be under-specified, and timings prone to rounding errors.)
+	 */
+	if (mode->vtotal == 405 || mode->vtotal == 819) {
+		/* Systems A and E (interlaced only) */
+		vec->fake_31khz = false;
 		mode_family = 1;
-	else if (tvstd == DRM_MODE_TV_MODE_PAL_M || tvstd == DRM_MODE_TV_MODE_PAL)
-		mode_family = 2;
-	else
-		mode_family = 0;
-	mode_narrow = (mode->clock >= 14336);
-	hwm = &rp1vec_hwmodes[mode_family][mode_ilaced][mode_narrow];
-	dev_info(&vec->pdev->dev,
-		 "%s: in_fmt=\'%c%c%c%c\' mode=%dx%d%s [%d%d%d] tvstd=%d",
-		__func__, in_format, in_format >> 8, in_format >> 16, in_format >> 24,
-		mode->hdisplay, mode->vdisplay, (mode_ilaced) ? "i" : "",
-		mode_family, mode_ilaced, mode_narrow, tvstd);
-
-	w = mode->hdisplay;
-	h = mode->vdisplay >> mode_ilaced;
-	if (w > hwm->total_cols)
-		w = hwm->total_cols;
-	if (h > hwm->rows_per_field)
-		h = hwm->rows_per_field;
+		hwm = &rp1vec_vintage_modes[(mode->vtotal == 819) ? 1 : 0];
+	} else {
+		/* 525- and 625-line modes, with half-height and "fake" progressive variants */
+		vec->fake_31khz = mode->vtotal >= 500 && !(mode->flags & DRM_MODE_FLAG_INTERLACE);
+		h = (mode->vtotal >= 500) ? (mode->vtotal >> 1) : mode->vtotal;
+		if (h >= 272)
+			mode_family = 1; /* PAL-625 */
+		else if (tvstd == DRM_MODE_TV_MODE_PAL_M || tvstd == DRM_MODE_TV_MODE_PAL)
+			mode_family = 2; /* PAL-525 */
+		else
+			mode_family = 0; /* NTSC-525 */
+		hwm = &rp1vec_hwmodes[mode_family][(mode->flags & DRM_MODE_FLAG_INTERLACE) ? 1 : 0];
+	}
 
 	/*
-	 * Add padding so a framebuffer with the given dimensions and
-	 * [hv]sync_start can be displayed in the chosen hardware mode.
-	 *
-	 *          |<----- mode->hsync_start ----->|
-	 *          |<------ w ------>|             |
-	 *          |                 |         >|--|<  ref_hfp
-	 *                            |<- hpad ->|
-	 * |<------------ total_cols ----------->|
-	 *  ________FRAMEBUFFERCONTENTS__________
-	 * '                                     `--\____/-<\/\/\>-'
+	 * Choose the upsampling rate (to 108MHz) in the range 4..16.
+	 * Clip dimensions to the limits of the chosen hardware mode, then add
+	 * padding as required, making some attempt to respect the DRM mode's
+	 * display position (relative to H and V sync start). Note that "wmax"
+	 * should be wider than the horizontal active region, to avoid boundary
+	 * artifacts (e.g. wmax = 728, w = 720, active ~= 704 in Rec.601 modes).
 	 */
-	hpad = max(0, mode->hsync_start - hwm->ref_hfp - w);
-	hpad = min(hpad, hwm->total_cols - w);
-	vpad = max(0, ((mode->vsync_start - hwm->ref_vfp) >> mode_ilaced) - h);
-	vpad = min(vpad, hwm->rows_per_field - h);
+	i = (vec->fake_31khz) ? (mode->clock >> 1) : mode->clock;
+	rate = (i < (RP1VEC_VDAC_KHZ / 16)) ? 16 : max(4, (RP1VEC_VDAC_KHZ + 256) / i);
+	wmax = min((hwm->de_end - hwm->de_bgn) / rate, 1020);
+	w = min(mode->hdisplay, wmax);
+	ref_2mid = (hwm->de_bgn + hwm->de_end) / rate + 6; /* + 6 for FIR delay */
```

Comment:
Could stand to reduce the magic offset to 4, shifting everything right by 1 pixel(!), to more precisely match Pi 4's video out (in Rec.601 modes at least).